### PR TITLE
TST: optimize.root_scalar: refactor tests and add Chandrupatla tests

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Supports gh-17719

#### What does this implement/fix?
This splits scalar root finder tests into separate classes for bracketing methods and gradient-based methods. It also reactors the bracketing method tests:
- Adds comments about what the tests do
- Eliminates some redundancy (copy-paste of functions, tolerances) and unused parameters
- Fixes some PEP8
- Uses `pytest.parametrize` instead of loops/separate tests

This also adds test cases from Chandrupatla's original paper.

### CODE REVIEW

File: scipy/optimize/tests/test_zeros.py [hunk 1]
```diff
@@ -18,12 +18,17 @@
...
+from scipy.optimize._tstutils import get_tests, functions as tstutils_functions
 
 TOL = 4*np.finfo(float).eps  # tolerance
 
 _FLOAT_EPS = finfo(float).eps
 
+bracket_methods = [zeros.bisect, zeros.ridder, zeros.brentq, zeros.brenth,
+                   zeros.toms748]
+gradient_methods = [zeros.newton]
+all_methods = bracket_methods + gradient_methods
```
Comment (mdhaber) [inline scipy/optimize/tests/test_zeros.py:30]:
```suggestion
all_methods = bracket_methods + gradient_methods  # type: ignore[operator]
```

**Comment by mdhaber** [comment]
@tupui would you be willing to skim this? i Much if it is a big, boring refactor but it will be useful for new development in `root_scalar` e.g. gh-17719.

**Review by tupui — APPROVED** [review]
LGTM, nothing jumping out and tests are green. Will let finish the mypy job and merge.

File: scipy/optimize/_tstutils.py [hunk 2]
```diff
@@ -674,3 +678,118 @@ def get_tests(collection='original', smoothness=None):
...
+    [fun9, [2e-4, 9], fun9.root, 10],
+    [fun9, [2e-4, 27], fun9.root, 11],
+    [fun9, [2e-4, 81], fun9.root, 13],
+]
+_CHANDRUPATLA_TESTS = [test + [f'{test[0].__name__}.{i%5+1}']
+                       for i, test in enumerate(_CHANDRUPATLA_TESTS)]
+
+_CHANDRUPATLA_TESTS_DICTS = [dict(zip(_CHANDRUPATLA_TESTS_KEYS, testcase))
+                             for testcase in _CHANDRUPATLA_TESTS]
+_add_a_b(_CHANDRUPATLA_TESTS_DICTS)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/optimize/_tstutils.py:795]: Yeah I thought about refactoring all that, too, but decided to limit the scope.

**Comment by mdhaber** [comment]
Oiy, mypy make up your mind!
@tupui  the original complaint was:

```
scipy/optimize/tests/test_zeros.py:30: error: Unsupported operand types for + ("List[function]" and "List[Callable[[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any], Any]]")  [operator]
```
So we added `  # type: ignore[operator]` on line 30 and now:
```
scipy/optimize/tests/test_zeros.py:30: error: Unused "type: ignore" comment
```
Not sure what to do!

**Comment by tupui** [comment]
Yup seeing that... Let's make it a full noqa then.

**Pull request merged**